### DISCUSSION TITLE
Modulators and their UI

### DISCUSSION THREAD

**Discussion by VegaDeftwing** [noisedeck/noisedeck-docs#14] [Ideas]
Right now the only 'modulator' accessible in Noisedeck is noise
![noise](https://user-images.githubusercontent.com/11283620/114896667-25f04e80-9dd6-11eb-9755-0d9d2ac3bd99.png)
Which I find to be less than useful. I can only assume there are plans to add more things here, but I'd like to suggest what I think would be the best way to do this

Clearly, any LFOs would have *some* limitations to keep the loop from being broken, so the LFO would need to have a rate that's a multiple of the length of the loop. The immediately obvious thing is to just implement the basic waveshapes (Sine, Upward saw, Downward saw, tri, square) and maybe shifts of each (variable phase sine, SAM/Super Saw, PWM etc.) but I think this could easily end up in a pretty gross UI **and** still be limiting.

So, that brings me to what I think would be best- a mod matrix with a mix of clever UI elements I've seen in VCV rack, starting with knocking out the wave shape problem:

![phase](https://user-images.githubusercontent.com/11283620/114899097-4ae5c100-9dd8-11eb-96c4-14a3be69c87d.png)
ZZC's VCV modules have this clever design for clock-synchronized LFO's by having everything be based on a main phasor, not a main clock tick. Here, I have a multiply/divider module being synced to that (in the middle) then that goes to a function generator that has the ability to do phase offset and a PWM-like operation. The PWM actually morphs from a downwald saw (as pictured) to a sine in the middle, to an upward saw when all the way right. If you just added a toggle to 'square' the wave to do 'real' PWM that would get almost all the waveshapes anybody could want. Only thing extra might be Sample and Hold... which honestly I'm not sure how you'd work in. 

Then, and this is the part I could see being arguable, I think a mod matrix would be great - but not for routing to the various operator, but rather just for making more complex ones.

![complexlfos](https://user-images.githubusercontent.com/11283620/114901203-30144c00-9dda-11eb-98a9-41286f5f362c.png)

So, I don't really think the output of earch 'lane' of the matrix should go directly to a particular slider in noisedeck, but instead just be a named modulator that can be selected as an alternative to noise or audio lo/mid/hi

From there, I think the current style of dialog box is a fine, maybe rename min to 'offset' and add an 'invert' toggle. 
![modulatoredit](https://user-images.githubusercontent.com/11283620/114901931-ea0bb800-9dda-11eb-8c7c-9e7636cd4005.png)

I have *no idea* how you'd work this into the current UI. What you have right now is super clean, and this is inherently complex, but I thought on it and I really can't come up with something better, and at least this doesn't require nested dialog or anything else super crazy, but I can see why this might be less than ideal if it meant you'd have to throw a completely new UI *somewhere* as below, left of, or right of the canvas are all a bit awkward, and above is wayy worse.

What do you two think?

**Comment by jesshewitt** [comment]
I definitely agree that our current audio/modulator support is very preliminary and not entirely useful. I really like the concept you've described here. We'll give it some thought, but I think it may be possible to implement something along these lines.
